Provide functionality to write custom ShardedTensor ops.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69874

We have a handful of ops supported for ShardedTensor via
``__torch_function__`` dispatch. However, we currently can't cover all torch
operators and having a way for users to extend this functionality will make
this functionality much more general.

In this PR, I've introduced a @custom_sharded_op decorator which can be used to
register a custom sharded op implementation.

Differential Revision: [D33078587](https://our.internmc.facebook.com/intern/diff/D33078587/)

pritamdamania committed:
Provide functionality to write custom ShardedTensor ops.

We have a handful of ops supported for ShardedTensor via
``__torch_function__`` dispatch. However, we currently can't cover all torch
operators and having a way for users to extend this functionality will make
this functionality much more general.

In this PR, I've introduced a @custom_sharded_op decorator which can be used to
register a custom sharded op implementation.

Differential Revision: [D33078587](https://our.internmc.facebook.com/intern/diff/D33078587/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -510,3 +512,38 @@ def shard_parameter(
+        >>> # This will call 'my_custom_sharded_linear'
+        >>> torch.nn.functional.linear(input, weight, bias)
+
+    The types, args and kwargs parameters are the same parameters that are
```
I just take a look on all the implementations we have, the `types` parameter seems not getting used anywhere, I am wondering if we should get rid of this argument both for custom op and for our own ops, `__torch_function__` can have it, but it seems like we don't need it for the actual computation?

pritamdamania87 replied:
I guess its safer to preserve types as it is provided to the user in `__torch_function__` as well.

wanchaol reviewed (approved):
looks good, only one small suggestion

fduwjj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -545,11 +559,13 @@ def sharding_spec(self) -> ShardingSpec:
 
     def __torch_function__(self, func, types, args=(), kwargs=None):
-        if func == torch.nn.functional.linear:
+        if func in _CUSTOM_SHARDED_OPS:
+            return _CUSTOM_SHARDED_OPS[func](types, args, kwargs, self._process_group)
+        elif func == torch.nn.functional.linear:
```
Do you feel we should always use this decorator for all new ops and migrate the existing ones using this decorator, too?

pritamdamania87 replied:
Yes, this indeed might be much more cleaner. Let me update the PR.

pritamdamania committed:
Update on "Provide functionality to write custom ShardedTensor ops."


We have a handful of ops supported for ShardedTensor via
``__torch_function__`` dispatch. However, we currently can't cover all torch
operators and having a way for users to extend this functionality will make
this functionality much more general.

In this PR, I've introduced a custom_sharded_op decorator which can be used to
register a custom sharded op implementation.

Differential Revision: [D33078587](https://our.internmc.facebook.com/intern/diff/D33078587/)

[ghstack-poisoned]

albanD reviewed (commented):
Thanks for the fix in nn.init, it looks good!

pritamdamania committed:
Update on "Provide functionality to write custom ShardedTensor ops."


We have a handful of ops supported for ShardedTensor via
``__torch_function__`` dispatch. However, we currently can't cover all torch
operators and having a way for users to extend this functionality will make
this functionality much more general.

In this PR, I've introduced a custom_sharded_op decorator which can be used to
register a custom sharded op implementation.

Differential Revision: [D33078587](https://our.internmc.facebook.com/intern/diff/D33078587/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Provide functionality to write custom ShardedTensor ops."


We have a handful of ops supported for ShardedTensor via
``__torch_function__`` dispatch. However, we currently can't cover all torch
operators and having a way for users to extend this functionality will make
this functionality much more general.

In this PR, I've introduced a custom_sharded_op decorator which can be used to
register a custom sharded op implementation.

Differential Revision: [D33078587](https://our.internmc.facebook.com/intern/diff/D33078587/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Provide functionality to write custom ShardedTensor ops."


We have a handful of ops supported for ShardedTensor via
``__torch_function__`` dispatch. However, we currently can't cover all torch
operators and having a way for users to extend this functionality will make
this functionality much more general.

In this PR, I've introduced a custom_sharded_op decorator which can be used to
register a custom sharded op implementation.

Differential Revision: [D33078587](https://our.internmc.facebook.com/intern/diff/D33078587/)

[ghstack-poisoned]
